scala/src/main/org/apache/spark/streaming/api/csharp/CSharpDStream.scala

```diff
@@ -343,3 +347,149 @@ class CSharpConstantInputDStream(ssc_ : StreamingContext, rdd: RDD[Array[Byte]])
+  private val pendingRddQueue: LinkedBlockingQueue[RddPreComputeRecord[T]] =
+    new LinkedBlockingQueue[RddPreComputeRecord[T]]()
+
+  private var jobExecutor = ThreadUtils.newDaemonFixedThreadPool(numThreads, s"job-executor-$id")
+  private val pendingQueueScheduler = ThreadUtils.newDaemonSingleThreadExecutor(
+    s"pending-scheduler-$id")
+
+  private def isPending(rddSeqNum: Long): Boolean = {
+    rddSeqNum > ackedSeqNum + maxUnackedNum
+  }
+
+  private def processPendingQueue(): Unit = {
+    while (true) {
+      val record = pendingRddQueue.take()
+      logInfo(s"process pending queue [$id], rddSeqNum: ${record.rddSeqNum}, " +
+        s"ackedSeqNum: $ackedSeqNum, outputRddSeqNum: $outputRddSeqNum")
+      while (isPending(record.rddSeqNum)) {
+        Thread.sleep(100)
+      }
+      logInfo(s"process pending queue [$id], submitting job, rddSeqNum: ${record.rddSeqNum}, " +
+        s"ackedSeqNum: $ackedSeqNum, outputRddSeqNum: $outputRddSeqNum")
+      jobExecutor.execute(new Runnable {
+        override def run(): Unit = {
+          record.rdd.persist(storageLevel)
+          sc.setCallSite(s"runJob for [$id]")
+          sc.runJob(record.rdd, (iterator: Iterator[T]) => {})
+        }
+      })
+    }
+  }
+
+  // input RDD
+  def slurp(userInfo: Any, rdd: RDD[T]): Unit = {
+    val rddSeqNum = inputRddSeqNum
+    inputRddSeqNum += 1
+    val record = RddPreComputeRecord[T](rddSeqNum, userInfo, rdd)
+    logInfo(s"slurp record for [$id], rddSeqNum: ${record.rddSeqNum}, " +
+      s"ackedSeqNum: $ackedSeqNum, outputRddSeqNum: $outputRddSeqNum")
+    if (isPending(record.rddSeqNum)) {
+      bypassRddMap.put(record.rddSeqNum, record)
+      pendingRddQueue.put(record)
+    } else {
+      logInfo(s"slurp, submitting job, rddSeqNum: ${record.rddSeqNum}, " +
+        s"ackedSeqNum: $ackedSeqNum, outputRddSeqNum: $outputRddSeqNum")
+      jobExecutor.execute(new Runnable {
+        override def run(): Unit = {
+          record.rdd.persist(storageLevel)
+          sc.setCallSite(s"runJob for [$id]")
+          sc.runJob(record.rdd, (iterator: Iterator[T]) => {})
+          precomputedRddMap.put(record.rddSeqNum, record)
+        }
+      })
+    }
+  }
+
+  // output RDDs
+  def spit(): List[RddPreComputeRecord[T]] = {
+    val result = new ListBuffer[RddPreComputeRecord[T]]()
+
+    var stop = false
+    var spitNum = 0
+    while (!stop) {
+      val rddSeqNum = outputRddSeqNum
+      if (precomputedRddMap.containsKey(rddSeqNum)) {
+        result += precomputedRddMap.remove(rddSeqNum)
+        outputRddSeqNum += 1
+        spitNum += 1
+        if (spitNum > spitLimit) {
+          stop = true
+        }
+      } else {
+        stop = true
+      }
+    }
+
+    if (result.isEmpty) {
+      stop = false
+      spitNum = 0
+      while (!stop) {
+        val rddSeqNum = outputRddSeqNum
+        if (bypassRddMap.containsKey(rddSeqNum)) {
+          result += bypassRddMap.remove(rddSeqNum)
```

Comment:
So, in this case, RDD is not precomputed right? (but it is still passed under the datastructure named RddPreComputeRecord?)